src/java/org/apache/cassandra/index/sai/disk/IndexSearchResultIterator.java

```diff
@@ -0,0 +1,141 @@
+     * Builds a new {@link IndexSearchResultIterator} that wraps a {@link KeyRangeUnionIterator} over the
+     * results of searching the {@link org.apache.cassandra.index.sai.memory.MemtableIndex} and the {@link SSTableIndex}es.
+     */
+    @SuppressWarnings({"resource", "RedundantSuppression"})
+    public static IndexSearchResultIterator build(Expression expression,
+                                                  Set<SSTableIndex> sstableIndexes,
+                                                  AbstractBounds<PartitionPosition> keyRange,
+                                                  QueryContext queryContext)
+    {
+        List<KeyRangeIterator> subIterators = new ArrayList<>(1 + sstableIndexes.size());
+
+        KeyRangeIterator memtableIterator = expression.context.getMemtableIndexManager().searchMemtableIndexes(expression, keyRange);
+        if (memtableIterator != null)
+            subIterators.add(memtableIterator);
+
+        for (SSTableIndex sstableIndex : sstableIndexes)
+        {
+            try
+            {
+                queryContext.checkpoint();
+                queryContext.incSstablesHit();
+                assert !sstableIndex.isReleased() : "Index was released from the view during the query";
+
+                SSTableQueryContext context = queryContext.getSSTableQueryContext(sstableIndex.getSSTable());
+                List<KeyRangeIterator> segmentIterators = sstableIndex.search(expression, keyRange, context);
+
+                if (!segmentIterators.isEmpty())
+                    subIterators.addAll(segmentIterators);
+            }
+            catch (Throwable e)
+            {
+                if (!(e instanceof QueryCancelledException))
+                    logger.debug(String.format("Failed search an index %s, aborting query.", sstableIndex.getSSTable()), e);
+
+                throw Throwables.cleaned(e);
+            }
+        }
+
+        KeyRangeIterator union = KeyRangeUnionIterator.build(subIterators);
+        return new IndexSearchResultIterator(union, sstableIndexes, queryContext);
+    }
+
+    protected PrimaryKey computeNext()
+    {
+        try
+        {
+            return union.hasNext() ? union.next() : endOfData();
+        }
+        finally
+        {
+            context.checkpoint();
+        }
+    }
+
+    protected void performSkipTo(PrimaryKey nextKey)
+    {
+        try
+        {
+            union.skipTo(nextKey);
+        }
+        finally
+        {
+            context.checkpoint();
+        }
+    }
+
+    public void close()
+    {
+        FileUtils.closeQuietly(union);
+        referencedIndexes.forEach(IndexSearchResultIterator::releaseQuietly);
+        referencedIndexes.clear();
+    }
+
+    private static void releaseQuietly(SSTableIndex index)
+    {
+        try
+        {
+            index.release();
+        }
+        catch (Throwable e)
+        {
+            logger.error(String.format("Failed to release index %s", index.getSSTable()), e);
```

Comment:
```suggestion
            logger.error(String.format(index.getIndexContext().logMessage("Failed to release index on SSTable %s."), index.getSSTable()), e);
```

Reply:
Ok, I hadn't read this one before commenting above but I have now wrapped all the informational messages with `logMessage`.